src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs

```diff
@@ -73,43 +80,49 @@ public class TreeRouter : IRouter
                 throw new ArgumentNullException(nameof(constraintLogger));
             }
 
-            _next = next;
             _trees = trees;
             _logger = routeLogger;
             _constraintLogger = constraintLogger;
 
-            var namedEntries = new Dictionary<string, TreeRouteLinkGenerationEntry>(
-                StringComparer.OrdinalIgnoreCase);
+            _namedEntries = new Dictionary<string, OutboundMatch>(StringComparer.OrdinalIgnoreCase);
+
+            var outboundMatches = new List<OutboundMatch>();
 
             foreach (var entry in linkGenerationEntries)
             {
+
+                var binder = new TemplateBinder(urlEncoder, objectPool, entry.RouteTemplate, entry.Defaults);
+                var outboundMatch = new OutboundMatch() { Entry = entry, TemplateBinder = binder };
+                outboundMatches.Add(outboundMatch);
+
                 // Skip unnamed entries
-                if (entry.Name == null)
+                if (entry.RouteName == null)
                 {
                     continue;
                 }
 
-                // We only need to keep one AttributeRouteLinkGenerationEntry per route template
+                // We only need to keep one OutboundMatch per route template
                 // so in case two entries have the same name and the same template we only keep
                 // the first entry.
-                TreeRouteLinkGenerationEntry namedEntry = null;
-                if (namedEntries.TryGetValue(entry.Name, out namedEntry) &&
-                    !namedEntry.Template.TemplateText.Equals(entry.Template.TemplateText, StringComparison.OrdinalIgnoreCase))
+                OutboundMatch namedMatch;
+                if (_namedEntries.TryGetValue(entry.RouteName, out namedMatch) &&
+                    !string.Equals(
+                        namedMatch.Entry.RouteTemplate.TemplateText,
+                        entry.RouteTemplate.TemplateText, 
+                        StringComparison.OrdinalIgnoreCase))
                 {
                     throw new ArgumentException(
-                        Resources.FormatAttributeRoute_DifferentLinkGenerationEntries_SameName(entry.Name),
+                        Resources.FormatAttributeRoute_DifferentLinkGenerationEntries_SameName(entry.RouteName),
                         nameof(linkGenerationEntries));
                 }
-                else if (namedEntry == null)
+                else if (namedMatch == null)
                 {
-                    namedEntries.Add(entry.Name, entry);
+                    _namedEntries.Add(entry.RouteName, outboundMatch);
                 }
             }
 
-            _namedEntries = namedEntries;
-
             // The decision tree will take care of ordering for these entries.
-            _linkGenerationTree = new LinkGenerationDecisionTree(linkGenerationEntries.ToArray());
+            _linkGenerationTree = new LinkGenerationDecisionTree(outboundMatches.ToArray());
```

Comment:
The decision tree wants it to be an array, I didn't see much reason to change this.